lib/clowne/utils/operation.rb

```diff
@@ -0,0 +1,63 @@
+# frozen_string_literal: true
+
+require 'clowne/utils/clone_mapper'
+
+module Clowne
+  module Utils
+    class Operation # :nodoc: all
+      THREAD_KEY = :"#{name}.clowne_operation"
+      DEFAULT_MAPPER = Utils::CloneMapper
+
+      private_constant :THREAD_KEY
+
+      class << self
+        def current
+          Thread.current[THREAD_KEY]
+        end
+
+        def wrap(operation_class: self, mapper: nil)
+          return yield if current
+
+          Thread.current[THREAD_KEY] = operation_class.new(mapper || DEFAULT_MAPPER.new)
+
+          current.tap do |operation|
+            operation.clone = yield
+            clear!
+          end
+        end
+
+        def clear!
+          Thread.current[THREAD_KEY] = nil
+        end
+      end
+
+      attr_accessor :clone
```

Comment:
`clone` could be confused with a `#clone` method, which present in many Ruby classes (especially, Rails); maybe, `clone_record` or `clone_result`?

Reply:
Totally agree, I already had problems with this lousy naming ... `clone_record` the best for me